src/java/org/apache/cassandra/net/async/SwappingByteBufDataOutputStreamPlus.java

```diff
@@ -0,0 +1,308 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.net.async;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.WritableByteChannel;
+
+import com.google.common.base.Function;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelPromise;
+import org.apache.cassandra.io.util.DataOutputStreamPlus;
+import org.apache.cassandra.io.util.Memory;
+import org.apache.cassandra.io.util.UnbufferedDataOutputStreamPlus;
+import org.apache.cassandra.utils.vint.VIntCoding;
+
+/**
+ * A {@link DataOutputStreamPlus} implementation that is backed by a {@link ByteBuf} that is swapped out with another when full.
+ * Each buffer has a {@link ChannelPromise} associated with it,
+ * so the downstream progress of buffer (as it is flushed) can be observed. Each message will be serialized into one
+ * or more buffers, and each of those buffer's promises are associated with the message. That way the message's promise
+ * and be assocaited with the promises of the buffers into which it was serialized. Several messages may be packed into
+ * the same buffer, so each of those messages will share the same promise.
+ */
+public class SwappingByteBufDataOutputStreamPlus extends DataOutputStreamPlus
+{
+    private static final Logger logger = LoggerFactory.getLogger(SwappingByteBufDataOutputStreamPlus.class);
+    private final Channel channel;
+    private final int bufferCapacity;
+
+    /**
+     * The current {@link ByteBuf} that is being written to.
+     */
+    private ByteBuf buf;
+
+    public SwappingByteBufDataOutputStreamPlus(Channel channel, int bufferCapacity)
+    {
+        this.channel = channel;
+        this.bufferCapacity = bufferCapacity;
+    }
+
+    public Channel channel()
+    {
+        return channel;
+    }
+
+    boolean checkBufSpace(int writeSize)
+    {
+        if (buf != null)
+        {
+            if (buf.writableBytes() >= writeSize)
+                return true;
+
+            channel.writeAndFlush(buf);
```

Comment:
Also same comment as before. Maybe you shouldn't flush until the caller calls flush? Since it's a stream buffering is expected.